src/core/lib/gprpp/time_util.cc

```diff
@@ -0,0 +1,81 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/time_util.h"
+
+namespace grpc_core {
+
+gpr_timespec ToGprTimeSpec(absl::Duration duration) {
+  if (duration == absl::InfiniteDuration()) {
+    return gpr_inf_future(GPR_TIMESPAN);
+  } else if (duration == -absl::InfiniteDuration()) {
+    return gpr_inf_past(GPR_TIMESPAN);
+  } else {
+    int64_t s = absl::IDivDuration(duration, absl::Seconds(1), &duration);
+    int64_t n = absl::IDivDuration(duration, absl::Nanoseconds(1), &duration);
+    return gpr_time_add(gpr_time_from_seconds(s, GPR_TIMESPAN),
```

Comment:
This is necessary because it has to have a canonical form. e.g. `gpr_timespec` should have nano which is >= 0 but `n` here can have a negative value. This might be done manually here for better performance but I think it's too invasive.

Reply:
Why would `n` be negative?  If it is, wouldn't `s` also be negative?  And if so, won't the resulting `gpr_timespec` have both fields negative?

I guess I don't understand in what case `gpr_time_add()` would do something different.

Reply:
For the case where `duration` is `absl::Nanoseconds(-10)`, `s` will be `0` and `n` will be `-10` but `gpr_timespec` will get `-1` for `tv_sec` and `999999990` for `tv_nsec`. It is handled by [this](https://github.com/grpc/grpc/blob/59c38f4e4daf88062b274c7a4f7bbd5db614ef3a/src/core/lib/gpr/time.cc#L79-L88).